data-pipeline-ffi/src/span.rs

```diff
@@ -0,0 +1,1009 @@
+pub unsafe extern "C" fn ddog_get_traces_size(ptr: *mut TracesBytes) -> usize {
+    if ptr.is_null() {
+        return 0;
+    }
+
+    let object = &mut *ptr;
+    object.len()
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_get_trace(ptr: *mut TracesBytes, index: usize) -> *mut TraceBytes {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let object = &mut *ptr;
+    if index >= object.len() {
+        return std::ptr::null_mut();
+    }
+
+    &mut object[index] as *mut TraceBytes
+}
+
+// ------------------ TraceBytes ------------------
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_traces_new_trace(ptr: *mut TracesBytes) -> *mut TraceBytes {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let object = &mut *ptr;
+    object.push(TraceBytes::default());
+    object.last_mut().unwrap_unchecked() as *mut TraceBytes
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_trace_new_span(ptr: *mut TraceBytes) -> *mut SpanBytes {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let object = &mut *ptr;
+    object.push(SpanBytes::default());
+    object.last_mut().unwrap_unchecked() as *mut SpanBytes
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_get_trace_size(ptr: *mut TraceBytes) -> usize {
+    if ptr.is_null() {
+        return 0;
+    }
+
+    let object = &mut *ptr;
+    object.len()
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_get_span(ptr: *mut TraceBytes, index: usize) -> *mut SpanBytes {
+    if ptr.is_null() {
+        return std::ptr::null_mut();
+    }
+
+    let object = &mut *ptr;
+    if index >= object.len() {
+        return std::ptr::null_mut();
+    }
+
+    &mut object[index] as *mut SpanBytes
+}
+
+// ------------------- SpanBytes -------------------
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_span_debug_log(ptr: *const SpanBytes) -> CharSlice<'static> {
+    if ptr.is_null() {
+        return CharSlice::empty();
+    }
+
+    let span = &*ptr;
+    let debug_str = format!("{:?}", span);
+
+    let boxed_str = debug_str.into_boxed_str();
+    let boxed_len = boxed_str.len();
+
+    let leaked_ptr = Box::into_raw(boxed_str) as *const c_char;
+
+    CharSlice::from_raw_parts(leaked_ptr, boxed_len)
+}
+
+#[no_mangle]
+#[allow(clippy::missing_safety_doc)]
+pub unsafe extern "C" fn ddog_free_charslice(slice: CharSlice<'static>) {
+    let data = slice.as_slice();
+
+    if data.is_empty() {
+        return;
+    }
+
+    let data_ptr = data.as_ptr() as *mut u8;
+
+    // Memory should be valid UTF-8 if it came from ddog_span_debug_log
+    drop(Box::from_raw(
+        std::str::from_utf8_mut(std::slice::from_raw_parts_mut(data_ptr, data.len()))
+            .unwrap_unchecked(),
+    ));
```

Comment:
Let's keep this simpler (a String is basically just a Vec anyway)
```suggestion
    drop(Vec::from_raw_parts(data_ptr, data.len(), data.len());
```